util/etcd/etcd.go

```diff
@@ -0,0 +1,391 @@
+)
+
+const etcdTimeout = 2 * time.Second
+
+// Exponential backoff for etcd operations
+var etcdBackoff = wait.Backoff{
+	Steps:    9,
+	Duration: 50 * time.Millisecond,
+	Factor:   2.0,
+	Jitter:   0.1,
+}
+
+type grpcDial = func(ctx context.Context, addr string) (net.Conn, error)
+
+// TODO: Write test client to fake etcd clusters for downstream consumers
+// client provides connection parameters for an etcd cluster
+type client struct {
+	dialerContext grpcDial
+	tlsConfig     *tls.Config
+	log           logr.Logger
+	timeout       time.Duration
+	backoffParams wait.Backoff
+	endpoint      string
+}
+
+// MemberAlarm represents an alarm type association with a cluster member
+type MemberAlarm struct {
+	// MemberID is the ID of the member associated with the raised alarm
+	MemberID uint64
+
+	// Condition is the type of condition which has been raised.
+	Type AlarmType
+}
+
+type AlarmType int32
+
+const (
+	// AlarmOK denotes that the cluster member is OK
+	AlarmOk AlarmType = 0
+	// AlarmNoSpace denotes that the cluster member has run out of disk space
+	AlarmNoSpace AlarmType = 1
+	// AlarmCorrupt denotes that the cluster member has corrupted data
+	AlarmCorrupt AlarmType = 2
+)
+
+// Taken from kubeadm
+// Member struct defines an etcd member; it is used for avoiding to spread
+// github.com/coreos/etcd dependencies
+type Member struct {
+	// ID is the ID of this cluster member
+	ID uint64
+	// Name is the human-readable name of the member. If the member is not started, the name will be an empty string.
+	Name string
+	// PeerURLs is the list of URLs the member exposes to the cluster for communication.
+	PeerURLs []string
+	// ClientURLs is the list of URLs the member exposes to clients for communication. If the member is not started, clientURLs will be empty.
+	ClientURLs []string
+	// IsLearner indicates if the member is raft learner.
+	IsLearner bool
+	// Alarms is the list of conditions for a member
+	Alarms []AlarmType
+}
+
+type Status struct {
+	// Leader is the member ID which the responding member believes is the current leader.
+	Leader uint64
+
+	// Version is the cluster protocol version used by the responding member
+	Version string
+
+	// Errors contains alarm/health information and status.
+	Errors []string
+
+	// IsLearner indicates if the member is raft learner.
+	IsLearner bool
+
+	// RaftIndex is the current raft committed index of the responding member.
+	RaftIndex uint64
+
+	// RaftTerm is the current raft term of the responding member.
+	RaftTerm uint64
+
+	// RaftAppliedIndex is the current raft applied index of the responding member.
+	RaftAppliedIndex uint64
+
+	// DbSizeInUse is the size of the backend database logically in use, in bytes, of the responding member.
+	DbSizeInUse int64
+
+	// DdSize is the size of the backend database physically allocated, in bytes, of the responding member.
+	DbSize int64
+}
+
+// Client defines the interface for etcd clients used by Cluster API
+type Client interface {
```

Comment:
We shouldn't force users of this library to depend on functions they don't need.

I know this is how Kubernetes does things but we don't have to follow their patterns. Please see Interface Segregation Principle here: https://dave.cheney.net/2016/08/20/solid-go-design

I would recommend making `client` public and returning the direct structs and letting the caller define an interface if they need to that specifies exactly the calls they need.

Reply:
Done